### ISSUE TITLE
Analyzer activated by EnforceCodeStyleInBuild csproj property not triggered when testing

### ISSUE THREAD

**Issue by Doraku** [dotnet/roslyn-sdk#1189]
Hello, I'm trying to create some suppressors for specific cases and I can't seem to find a way for the expected diagnostic to trigger in my test.
In short I want to create a suppressor for [IDE0130](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/style-rules/ide0130) when I detect an type with extension methods as I like the extension type to have the same namespace as the extended type.

This is my tester code
```cs
﻿using System.Collections.Generic;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.CSharp.Testing;
using Microsoft.CodeAnalysis.Diagnostics;
using Microsoft.CodeAnalysis.Testing;

namespace DefaultCSharp;

public sealed class CSharpAnalyzerTest<TAnalyzer> : CSharpCodeFixTest<TAnalyzer, EmptyCodeFixProvider, DefaultVerifier>
        where TAnalyzer : DiagnosticAnalyzer, new()
{
    private readonly List<DiagnosticAnalyzer> _additionalDiagnosticAnalyzers = [];

    protected override CompilationOptions CreateCompilationOptions() => base.CreateCompilationOptions().WithGeneralDiagnosticOption(ReportDiagnostic.Warn);

    protected override IEnumerable<DiagnosticAnalyzer> GetDiagnosticAnalyzers() => [.. base.GetDiagnosticAnalyzers(), .. _additionalDiagnosticAnalyzers];

    public CSharpAnalyzerTest<TAnalyzer> AddSource(string code)
    {
        TestState.Sources.Add(code);

        return this;
    }

    public CSharpAnalyzerTest<TAnalyzer> AddAnalyzers(params DiagnosticAnalyzer[] analyzers)
    {
        _additionalDiagnosticAnalyzers.AddRange(analyzers);

        return this;
    }

    public CSharpAnalyzerTest<TAnalyzer> AddExpectedDiagnostics(params DiagnosticResult[] diagnostics)
    {
        TestState.ExpectedDiagnostics.AddRange(diagnostics);

        return this;
    }
}
```

and its usage
```cs
using System.Threading.Tasks;
using Microsoft.CodeAnalysis;
using Microsoft.CodeAnalysis.Testing;
using Xunit;

namespace DefaultCSharp.Supressors;

public sealed class IDE0130SupressorShould
{
    [Fact]
    public Task SuppressIDE0130WhenContainsExtensionsType()
    {
        return Task.CompletedTask;
        return new CSharpAnalyzerTest<IDE0130Supressor>()
            .AddSource(
                /* lang=c#-test */ """
                {|#0:namespace System;|}

                internal static class StringExtensions
                {
                    public static void Dummy(this string value)
                    { }
                }
                """)
            .AddAnalyzers(
                MicrosoftCodeAnalysisCSharpCodeStyleAnalyzers.IDE0130())
            .AddExpectedDiagnostics(
                new DiagnosticResult("IDE0130", DiagnosticSeverity.Warning)
                    .WithLocation(0)
                    .WithIsSuppressed(true))
            .RunAsync();
    }
}
```

The analyzer for IDE0130 is retrived like this as the type is internal, I verified in debug that is it correctly retrieved and instanciated
```cs
﻿using System;
using System.IO;
using System.Reflection;
using System.Runtime.InteropServices;
using System.Runtime.Loader;
using Microsoft.CodeAnalysis.Diagnostics;

namespace DefaultCSharp;

public class AnalyzerLoaderBase
{
    private static readonly Lazy<string> _nuGetPackagesFolder = new(GetNuGetPackagesFolder, isThreadSafe: true);

    protected static string NuGetPackagesFolder => _nuGetPackagesFolder.Value;

    protected static Type FindType(
        Lazy<Assembly> assembly,
        string typeName) =>
            assembly.Value.GetType(typeName) ?? throw new InvalidOperationException($"Could not locate type '{typeName}' from '{assembly.Value.GetName().Name}'");

    private static string GetNuGetPackagesFolder()
    {
        string result = Environment.GetEnvironmentVariable("NUGET_PACKAGES");
        if (result is null)
        {
            string homeFolder =
                RuntimeInformation.IsOSPlatform(OSPlatform.Windows)
                    ? Environment.GetEnvironmentVariable("USERPROFILE")
                    : Environment.GetEnvironmentVariable("HOME");

            result = Path.Combine(homeFolder ?? throw new InvalidOperationException("Could not determine home folder"), ".nuget", "packages");
        }

        if (!Directory.Exists(result))
        {
            throw new InvalidOperationException($"NuGet package cache folder '{result}' does not exist");
        }

        return result;
    }

    protected static Assembly LoadAssembly(string assemblyPath) => AssemblyLoadContext.Default.LoadFromAssemblyPath(assemblyPath);
}

public class MicrosoftCodeAnalysisCSharpCodeStyleAnalyzers : AnalyzerLoaderBase
{
    private static readonly Lazy<Assembly> _assemblyCSharpNetAnalyzers = new(LoadCSharpNetAnalyzers, isThreadSafe: true);
    private static readonly Lazy<Type> _typeIDE0130 = new(() => FindType(_assemblyCSharpNetAnalyzers, "Microsoft.CodeAnalysis.CSharp.Analyzers.MatchFolderAndNamespace.CSharpMatchFolderAndNamespaceDiagnosticAnalyzer"), isThreadSafe: true);

    public static DiagnosticAnalyzer IDE0130() =>
        Activator.CreateInstance(_typeIDE0130.Value) as DiagnosticAnalyzer ?? throw new InvalidOperationException($"Could not create instance of '{_typeIDE0130.Value.FullName}'");

    private static Assembly LoadCSharpNetAnalyzers()
    {
        LoadAssembly(Path.Combine(NuGetPackagesFolder, "microsoft.codeanalysis.csharp.codestyle", "4.11.0", "analyzers", "dotnet", "cs", "Microsoft.CodeAnalysis.CodeStyle.dll"));

        return LoadAssembly(Path.Combine(NuGetPackagesFolder, "microsoft.codeanalysis.csharp.codestyle", "4.11.0", "analyzers", "dotnet", "cs", "Microsoft.CodeAnalysis.CSharp.CodeStyle.dll"));
    }
}
```

I tried giving an editorconfig to my `CSharpAnalyzerTest` to enable all analyzers and the namespace rule like so
```
dotnet_analyzer_diagnostic.severity = warning
dotnet_style_namespace_match_folder = true
```
but it didn't change anything. In the build, those diagnostics are only reported if the `EnforceCodeStyleInBuild` property is set to true so I'm wondering if this is not what is missing but I can't find how to properly pass this into the compilation configuration of the `TestState` (if it's even the issue :/). Sorry for the code dump but I don't see how to explain it easier ^^". Hopefully my issue ring a bell to someone as for now I'm just skimming the code. I guess my next step would be to duplicate the `CSharpMatchFolderAndNamespaceDiagnosticAnalyzer` so I can properly debug it and see why it doesn't produce a diagnostic if all fails.

My test project is available [here](https://github.com/Doraku/DefaultCSharp/tree/master/source/DefaultCSharp.Tests) just in case

**Comment by sharwell** [comment]
When debugging the test, have you reviewed each exception thrown during the test to see if it's interfering with the behavior?

IDE0130 is dependent on the location of files in the project, so I'm not sure if/how the test code would be accounting for that.

**Comment by Doraku** [comment]
I'm not seeing any exception, the code does produce some build diagnostics because of missing xml documentation for some reason but they are filtered before the comparison to the expected ones.
Internally I've seen that adding source to the TestState automatically use numbered file name like "test#.cs" so if it actually execute it should see the missmatch (?). When I can I'll give the infos I can gather when I deep dive in debug, I remember seeing some counters of element processed somewhere and namespace wasn't at 0 🤔

**Comment by Doraku** [comment]
So because I'm stuck in commute/train I did some searching, which honestly I should have done first before opening this issue ^^", there is some test on the roslyn repo to validate this analyzer, it probably have to do with changing the default namespace of the solution like they do here
https://github.com/dotnet/roslyn/blob/55a809f91cae01bf1e6c635ee91faae6b0af23fe/src/Analyzers/CSharp/Tests/MatchFolderAndNamespace/CSharpMatchFolderAndNamespaceTests.cs#L46

Will report my findings when I try it.

**Comment by Doraku** [comment]
I learned that you can setup build property via an editor config file in the test like so:
```
TestState.AnalyzerConfigFiles.Add((
    "/0/.editorconfig",
    """
    is_global=true
    build_property.ProjectDir = /0/
    build_property.RootNamespace = Test
    """));
```

The diagnostic is now correctly produced in my test and I can move into actually implementing the suppressor :) Sorry for the rubberducking issue!
